### PR TITLE
light version of #17938: fix most issues with UnusedImport, XDeclaredButNotUsed, etc;  fix #17511, #17510, #14246

### LINKED ISSUE #14246: Incorrect `UnusedImport`  warning with re-exported symbols or with  `tyAlias` (`type Foo* = Bar`)
### Example
```nim
import strutils as su
export su
```

### Current Output
```
Warning: imported and not used: 'strutils' [UnusedImport]
```
Ps:
```nim
import strutils
export strutils
```
Don't have this warnning.

```
$ nim -v
Nim Compiler Version 1.2.0
```

### LINKED ISSUE #17510: Incorrect `UnusedImport` warnings regression from 1.4.2 to 1.5.1
When one module imports and exports a module, another module depending on the imported/exported module claims in 1.5.1 that the intermediate `import`/`export`ing module is unused.

### Example
`test.nim`:
```nim
import sequtils
export sequtils
```
`test2.nim`:
```nim
import "."/test
echo repeat("b", 3)
```

### Current Output
With head:
```
$ ~/Nim/bin/nim c -r test2.nim
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
.......
test2.nim(1, 11) Warning: imported and not used: 'test' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: test2.nim
Hint:  [Link]
Hint: 25592 lines; 0.544s; 31.934MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ ~/Nim/bin/nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 8573160a44439ffe3fbbdd354188db9c976a2081
active boot switches: -d:release
```

Nim 1.4.2 does not appear to emit this incorrect warning:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
.......CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim

Hint:  [Link]
Hint: 25242 lines; 0.534s; 25.52MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-12-02
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
$ 
```
Nor does 1.0.4.

### Expected Output
No incorrect `UnusedImport` warning

### LINKED ISSUE #17511: Incorrect `UnusedImport` warning in presence of `import foo as bar`
When module `bar` imports and exports module `foo` (here, `sequtils`) using `import ... as ...` and module `baz` imports `bar` to get access to `foo`, either one or two `UnusedImport` warning is emitted, depending on Nim version.

### Example
`test.nim`:
```nim
import sequtils as su2
export su2
```

`test2.nim`:
```nim
import "."/test
echo repeat("b", 3)
```

### Current Output
```
$ ~/Nim/bin/nim c -r test2.nim
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
.......
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
test2.nim(1, 11) Warning: imported and not used: 'test' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: test2.nim
Hint:  [Link]
Hint: 25592 lines; 0.530s; 31.934MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
user@tertiva:~$ ~/Nim/bin/nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 8573160a44439ffe3fbbdd354188db9c976a2081
active boot switches: -d:release
```

Older versions also display variations of this behavior.
1.4.2:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
.......
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: 25242 lines; 0.542s; 25.52MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-12-02
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

1.2.6:
```
$ Nim1.2.6/bin/nim c -r test2.nim
Hint: used config file 'Nim1.2.6/config/nim.cfg' [Conf]
Hint: used config file 'Nim1.2.6/config/config.nims' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: test2 [Processing]
Hint: test [Processing]
Hint: sequtils [Processing]
Hint: macros [Processing]
/home/user/test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: 25264 LOC; 0.532 sec; 30.262MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ Nim1.2.6/bin/nim --version
Nim Compiler Version 1.2.6 [Linux: amd64]
Compiled at 2021-03-24
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: f05ba3c667157f46f141ff700121c7aff86cf07a
active boot switches: -d:release
```

1.0.4:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: test2 [Processing]
Hint: test [Processing]
Hint: sequtils [Processing]
Hint: macros [Processing]
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: operation successful (25045 lines compiled; 0.523 sec total; 29.773MiB peakmem; Debug Build) [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.0.4 [Linux: amd64]
Compiled at 2020-02-04
Copyright (c) 2006-2019 by Andreas Rumpf

active boot switches: -d:release
```

### Expected Output
No incorrect `UnusedImport` warning(s).

### Additional Information
https://github.com/nim-lang/Nim/issues/17510 is related, but only appears after Nim 1.4.2 and doesn't need `import foo as bar` / `export bar`, just `import foo` / `export foo`.

### PR DESCRIPTION
**Pull request by timotheecour**
refs https://github.com/nim-lang/Nim/pull/17938#issuecomment-833890135, this PR is a light weight version of https://github.com/nim-lang/Nim/pull/17938, removing these:
* `{.used: symbol.}`
* the fix for #13185
* `genPNode`

ie, it does the following:

* module imports are now consistently creating a module alias for all forms of imports, eg `import foo` vs `import foo as foo2` both call `createModuleAliasImpl`; this allows keeping module modifications  (eg `moduleSym.flags`) isolated from the actual imported module
* fix #17511, see [1]
* add new warning `DuplicateModuleImport` that triggers for eg on `import foo; import foo`
* fix a bug where `import foo as bar` would not show deprecation msg for foo (before PR, only `import foo` would show the deprecation msg), eg with `{.deprecated: "blah".}`
* fix #17510
* fix #14246
* relies on the recent `nimoutFull` which is critical for this kind of test

## note
* [1] #17511 has 2 components A and B, I'm fixing both:
A is the `test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]` (warning in the imported module referring to:
```nim
import sequtils as su2
export su2
```
B is the 2nd warning `test2.nim(1, 11) Warning: imported and not used: 'test' [UnusedImport]` which is a regression

### CODE REVIEW

File: compiler/ast.nim [hunk 1]
```diff
@@ -853,6 +853,8 @@ type
   TSym* {.acyclic.} = object of TIdObj # Keep in sync with PackedSym
     # proc and type instantiations are cached in the generic symbol
     case kind*: TSymKind
+    of skModule:
+      realModule*: PSym # for `createModuleAlias`
```
Comment (Araq) [inline compiler/ast.nim:857]: I know you want me to review and accept it, but I'm still thinking about the implications of this data model extension.

**Comment by Varriount** [comment]
Sooner or later, I think it might be prudent to have a `module` type (even if only as an input to RTTI procedures).

**Comment by Araq** [comment]
Here is how it should be done instead: For `import m` the m PSym is **copied** and this copy is inserted into the importing module's symbol table. Since it's always a copy, it can also have a different name so `import m as m2` has a most natural representation.

**Comment by Araq** [comment]
> Sooner or later, I think it might be prudent to have a module type (even if only as an input to RTTI procedures).

We already have an `ImportedModule` type in the compiler...

**Comment by timotheecour** [comment]
PTAL, rebased after bitrot conflicts

> Here is how it should be done instead: For import m the m PSym is copied and this copy is inserted into the importing module's symbol table. Since it's always a copy, it can also have a different name so import m as m2 has a most natural representation.

I'm already doing that, createModuleAlias is always called for imports and creates a PSym copy

EDIT:
it was green before rebase but after rebase, CI now fails for IC with following error:
```
Hint:  [Link]
duplicate symbol '_g_systemZschubfach_81' in:
    /Users/timothee/git_clone/nim/Nim_prs/nimcache/tests/ic/timports.nim_4a8a08f09d37b73795649038408b5f33/stdlib_system.nim.c.o
    /Users/timothee/git_clone/nim/Nim_prs/nimcache/tests/ic/timports.nim_4a8a08f09d37b73795649038408b5f33/stdlib_schubfach.nim.c.o
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
(the other errors are the same)

File: compiler/ast.nim (hunk 1 above)
Comment (Araq) [reply compiler/ast.nim:857]: This is not required and harmful for IC.

**Comment by timotheecour** [comment]
PTAL: it turns out the `duplicate symbol` link error is caused by a pre-existing bug in IC (just filed https://github.com/nim-lang/Nim/issues/18361 which explains it) that wasn't exhibited until this PR because until this PR, `createModuleAlias` wasn't being called for `import foo` (only for `import foo as bar`)

**Comment by timotheecour** [comment]
trying another PR without realModule here: https://github.com/nim-lang/Nim/pull/18362

File: compiler/ast.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/ast.nim:857]: => https://github.com/nim-lang/Nim/pull/18362

**Pull request closed without merging**